agent/src/bpf/audit.bpf.c

```diff
@@ -226,6 +234,126 @@ record_blob_data (struct pt_regs *ctx, long context, const char *key_ptr)
   return err;
 }
 
+/* This is similar to record_blob_data but always assumes VALUE_PTR is
+ * present.
+ */
+static __always_inline int
+record_blob_data_explicit (struct pt_regs *ctx, long context, const char *key_ptr,
+			   const void *value_ptr, long value_size)
+{
+  int err;
+
+  struct audit_blob_data_event_st *event =
+    bpf_ringbuf_reserve (&ringbuf,
+			 sizeof(struct audit_blob_data_event_st),
+			 0);
+  if (!event)
+    {
+      DEBUG ("unable to allocate ringbuf entry: %ld\n", -ENOMEM);
+      return -ENOMEM;
+    }
+
+  populate_event_header (&event->base.header,
+			 AUDIT_EVENT_DATA,
+			 sizeof(*event),
+			 context);
+
+  event->base.type = AUDIT_DATA_BLOB;
+  err = bpf_probe_read_user_str (event->base.key, KEY_SIZE, (void *)key_ptr);
+  if (err < 0)
+    {
+      DEBUG ("unable to read event key: %ld\n", err);
+      goto error;
+    }
+
+  if (value_size > 0)
+    {
+      value_size &= (VALUE_SIZE - 1);
+      err = bpf_probe_read_user (event->value, value_size, (void *)value_ptr);
+      if (err < 0)
+	{
+	  DEBUG ("unable to read event data: %ld\n", err);
+	  goto error;
+	}
+    }
+
+  event->size = value_size;
+
+  bpf_ringbuf_submit (event, 0);
+  return 0;
+
+ error:
+  bpf_ringbuf_discard (event, 0);
+  return err;
+}
+
+static __always_inline int
+record_data (struct pt_regs *ctx,
+	     long context,
+	     const struct crypto_auditing_data *array_ptr,
+	     long array_size)
+{
+  struct crypto_auditing_data events[MAX_EVENTS];
+  size_t events_size = array_size & (MAX_EVENTS - 1);
+  int err = 0;
+
+  err = bpf_probe_read_user (events,
+			     events_size * sizeof (struct crypto_auditing_data),
+			     array_ptr);
+  if (err < 0)
+    {
+      DEBUG ("unable to read from data array: %ld\n", err);
+      return err;
+    }
+
+  for (long i = 0; i < events_size; i++)
+    {
+      switch (events[i].value_size)
```

Comment:
is it worth it separating an overloaded value_size into type + unsigned size?

Reply:
I initially thought so, but decided to consolidate them as we will not add more types. There are macros `CRYPTO_AUDITING_STRING`, `CRYPTO_AUDITING_WORD`, and `CRYPTO_AUDITING_BLOB` provided to support writing the entries easily.

Reply:
OK, as long as we don't need more variable-sized types.